dbms/src/DataStreams/LiveViewBlockInputStream.h

```diff
@@ -0,0 +1,222 @@
+/* Copyright (c) 2018 BlackBerry Limited
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+#pragma once
+
+#include <limits>
+
+#include <Common/ConcurrentBoundedQueue.h>
+#include <Poco/Condition.h>
+#include <DataStreams/OneBlockInputStream.h>
+#include <DataStreams/IBlockInputStream.h>
+#include <Storages/StorageLiveView.h>
+
+
+namespace DB
+{
+
+/** Implements LIVE VIEW table WATCH input stream.
+ *  Keeps stream alive by outputing blocks with no rows
+ *  based on period specified by the heartbeat interval.
+ */
+class LiveViewBlockInputStream : public IBlockInputStream
+{
+
+using NonBlockingResult = std::pair<Block, bool>;
+
+public:
+    ~LiveViewBlockInputStream() override
+    {
+        /// Start storage no users thread
+        /// if we are the last active user
+        if (!storage->is_dropped && blocks_ptr.use_count() < 3)
```

Comment:
I think it's not very safe to check use_count.
If somebody else will copy blocks_ptr, no thread will be created.

Also, code is not thread-safe. Two threads which concurrently call `~LiveViewBlockInputStream()` could see that condition `blocks_ptr.use_count() < 3` is not true, and only after that remove their ptrs. So, `storage->startNoUsersThread` won't be called.

I think the good option to solve this is to use manual atomic use counter. You may create class like 
```c++
class SharedBlocks
{
private:
    BlocksPtr blocks;
    std::atomic<size_t> use_count;
}
``` 
and RAII guard class that will anatomically decrease use_count in destructor and call something if it is 0.

Or maybe use custom deleter for shared_ptr. Don't know yet if it can help.

Reply:
Also, if you add `mutex` and `condition` in `SharedBlocks`, it will probably be possible not to store `storage` field.